File: server/nodejs/controllers/userController.js
const express = require('express');
const router = express.Router();
const { User, UserRole } = require('../models/user');

// Mendapatkan semua pengguna
router.get('/users', (req, res) => {
  User.getAllUsers((err, users) => {
    if (err) {
      console.error('Error fetching users:', err);
      res.status(500).json({ error: 'Server error' });
    } else {
      res.json(users);
    }
  });
});

// Mendapatkan pengguna berdasarkan ID
router.get('/users/:id', (req, res) => {
  const userId = req.params.id;
  User.getUserById(userId, (err, user) => {
    if (err) {
      console.error('Error fetching user:', err);
      res.status(500).json({ error: 'Server error' });
    } else if (!user) {
      res.status(404).json({ message: 'User not found' });
    } else {
      res.json(user);
    }
  });
});

// Membuat pengguna baru
router.post('/user/create', (req, res) => {
  const { username, password, role, nama_pelanggan, alamat, no_telp } = req.body;

  User.createUser(username, password, role, nama_pelanggan, alamat, no_telp, (err, result) => {
    if (err) {
      console.error('Error creating user:', err);
      res.status(500).json({ error: 'Server error' });
    } else {
      res.status(201).json({ message: 'User created', user: { id: result.insertId, username, role, nama_pelanggan, alamat, no_telp } });
    }
  });
});

// Memperbarui pengguna berdasarkan ID
router.put('/user/update/:id', (req, res) => {
  const userId = req.params.id;
  const { username, password,role, nama_pelanggan, alamat, no_telp } = req.body;
  User.updateUser(userId, username, password, role, nama_pelanggan, alamat, no_telp, (err, result) => {
    if (err) {
      console.error('Error updating user:', err);
      res.status(500).json({ error: 'Server error' });
    } else if (result.affectedRows === 0) {
      res.status(404).json({ message: 'User not found' });
    } else {
      res.json({ message: 'User updated', userId });
    }
  });
});

// Menghapus pengguna berdasarkan ID
router.delete('/user/delete/:id', (req, res) => {
  const userId = req.params.id;
  User.deleteUser(userId, (err, result) => {
    if (err) {
      console.error('Error deleting user:', err);
      res.status(500).json({ error: 'Server error' });
    } else if (result.affectedRows === 0) {
      res.status(404).json({ message: 'User not found' });
    } else {
      res.json({ message: 'User deleted', userId });
    }
  });
});


module.exports = router;

File: server/dotnet/Stackup.Api/Controllers/UserController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Configuration;
using System.Data;

[Route("api/[controller]")]
[ApiController]
public class UserController : Controller
{
    private readonly DbManager _dbManager;
    Response response = new Response();

    public UserController(IConfiguration configuration)
    {
        _dbManager = new DbManager(configuration);
    }

    // GET: api/User
    [HttpGet]
    public IActionResult GetUsers()
    {
        try
        {
            response.status = 200;
            response.message = "Succes";
            response.data = _dbManager.GetAllUsers();
        }
        catch (Exception ex)
        {
            response.status = 500;
            response.message = ex.Message;
        }
        return Ok(response);
    }

    // CREATE
    [HttpPost("/User/create")]
    public IActionResult CreateUser([FromBody]User user)
    {
        try
        {
            // Memeriksa role pengguna
            if (user.role != "admin" && user.role != "customer")
            {
                response.status = 403;
                response.message = "Unauthorized role";
                return StatusCode(403, response);
            }

            // Jika role sesuai, lanjutkan dengan membuat pengguna
            System.Diagnostics.Debug.WriteLine(user); //ini console log
            response.status = 200;
            response.message = "Success";
            _dbManager.CreateUser(user);
        }
        catch (Exception ex)
        {
            response.status = 500;
            response.message = ex.Message;
        }

        return Ok(response);
    }

    [HttpGet]
    [Route("/User/Usernames")] // Use a descriptive route
    public IActionResult GetUsernames()
    {
        try
        {
            List<string> usernames = _dbManager.GetUsernames();
            return Ok(usernames); // Return list of usernames as JSON
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.Message);
            return StatusCode(500, "Internal Server Error"); // Handle errors gracefully
        }
    }

    
    // Login
    [HttpGet("/User/login")]
    public IActionResult Login(string username, string password)
    {
        try
        {
            User user = _dbManager.Login(username, password);
            if (user == null)
            {
                return Unauthorized("Invalid username or password");
            }
            return Ok(new { id=user.id, role = user.role, nama_pelanggan = user.nama_pelanggan, alamat = user.alamat, no_telp = user.no_telp });
        }
        catch (Exception ex)
        {
            return StatusCode(500, ex.Message);
        }
    }


    // UPDATE
    [HttpPut("/User/update/{id}")]
    public IActionResult UpdateUser(int id, [FromBody]User user)
    {
        try
        {
            response.status = 200;
            response.message = "Success";
            _dbManager.UpdateUser(id, user);
        }
        catch(Exception ex)
        {
            response.status = 500;
            response.message = ex.Message;
        }
        return Ok(response);
    }

    // DELETE
    [HttpDelete("/User/delete/{id}")]
    public IActionResult DeleteUser(int id)
    {
        try
        {
            response.status = 200;
            response.message = "Success";
            _dbManager.DeleteUser(id);
        }
        catch(Exception ex)
        {
            response.status = 500;
            response.message = ex.Message;
        }
        return Ok(response);
    }
}